wordpress-mobile/WordPress-iOS
# Blogging Prompts: Migrate to v3 logic

Refs https://github.com/wordpress-mobile/WordPress-iOS/issues/20496

> **Note**
> Originally, this was meant to be done partially in WordPressKit (see https://github.com/wordpress-mobile/WordPressKit-iOS/pull/592), but I've decided to port the logic here since the Jetpack app is only the sole consumer of Blogging Prompts. 
> 
> See https://github.com/wordpress-mobile/WordPressKit-iOS/pull/592#issuecomment-1788042166 for more context. 

This PR migrates the Blogging Prompts fetching mechanism to the v3 version. Just so you know, the new endpoint version **will return a different set of prompts** when compared to v2, despite having the same date range. Some other notes:

- With the fetching logic now ported to `WordPress-iOS`, we now use `BloggingPromptRemoteObject` to represent a blogging prompt object from the v3 API.
  - I didn't port the Blogging Prompt Settings logic just yet to keep things manageable. It's a low priority, but once it's ported, we can remove the dependency on `BloggingPromptServiceRemote`.
- The `upsert` method is updated to overwrite based on `date` instead of `promptID`. Furthermore, the method ensures that there will only be one prompt per date. When processing new prompts, the first matching existing prompt will be overwritten, and the remaining "duplicates" will be deleted.
- Adjusted the unit tests and added new ones for various upsert scenarios.
- For locally scheduled prompts, any pending prompt reminders should be rescheduled and updated on app launch. See: [BloggingPromptCoordinator#updatePromptsIfNeeded](https://github.com/wordpress-mobile/WordPress-iOS/blob/901e6bf5ec58e043ee02e4529a88f45e2b5f61c8/WordPress/Classes/ViewRelated/Blog/Blogging%20Prompts/BloggingPromptCoordinator.swift#L82) and [PromptRemindersScheduler#processSchedule](https://github.com/wordpress-mobile/WordPress-iOS/blob/901e6bf5ec58e043ee02e4529a88f45e2b5f61c8/WordPress/Classes/Utility/Blogging%20Prompts/PromptRemindersScheduler.swift#L148)
- With this update, the `title` and `content` properties in `BloggingPrompt` are now unused. I'll make a separate PR that cleans up the property from the Core Data model to keep this PR manageable.

## To test

> **Important**
Before testing this PR, launch a previous version of the Jetpack app to locally save some prompts from the v2 API. You can do this by going to the prompts card > ellipsis menu > "View more prompts" to load v2 prompts about 10 days back.

- Launch the Jetpack app.
- Log in to a WP.com account.
- 🔎 Verify that the blogging prompt card appears on the home dashboard.
- Tap "View all responses".
- 🔎 Verify that the Reader stream shows the posts related to the prompt.
- Go back and tap "Answer prompt".
- 🔎 Verify that the draft post is opened and shows the prompt content in quotes.
- Go back, tap the ellipsis menu, and select "View more prompts".
- 🔎 Verify that the lists are loaded correctly, and no duplicated dates are shown.

#### Prompt scheduling

- Install the app store version of Jetpack.
- Schedule a prompt.
- Update the Jetpack with this version and launch the app.
- Wait for the reminder to be fired.
- 🔎 Verify that the reminder shows the contents of the new v3 prompts.

## Regression Notes
1. Potential unintended areas of impact
Prompts may be incorrectly loaded, or there could be other unexpected behavior with the Blogging Prompt features.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manually tested the changes, as described in the testing steps.

3. What automated tests I added (or what prevented me from doing so)
Adjusted the unit tests and added some new ones.

PR submission checklist:


UI Changes testing checklist:

N/A
